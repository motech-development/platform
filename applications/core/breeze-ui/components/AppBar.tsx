import { createElement, FC, ReactNode } from 'react';

type TElement = 'header' | 'div' | 'nav';

type TColour = 'primary' | 'secondary';

const parentStyles = (colour: TColour, fixed: boolean) => {
  let colourStyles: string;
  let positionStyles: string;

  if (colour === 'primary') {
    colourStyles = 'bg-gray-900 text-gray-100';
  } else {
    colourStyles = 'bg-gray-200 text-gray-700';
  }

  if (fixed) {
    positionStyles = 'fixed';
  } else {
    positionStyles = 'relative';
  }

  return `${colourStyles} ${positionStyles}`;
};

const innerStyles = (colour: TColour) =>
  colour === 'primary' ? 'border-gray-800' : 'border-gray-300';

const Parent: FC<{
  children: ReactNode;
  colour: TColour;
  element: TElement;
  fixed: boolean;
}> = ({ children, colour, element, fixed }) =>
  createElement(
    element,
    {
      className: parentStyles(colour, fixed),
    },
    children,
  );

const AppBar: FC<{
  children: ReactNode;
  colour?: TColour;
  element?: TElement;
  fixed?: boolean;
}> = ({ children, colour = 'primary', element = 'header', fixed = false }) => {
  return (
    <Parent colour={colour} element={element} fixed={fixed}>
      <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8">
        <div
          className={`flex items-center justify-between h-16 border-b-2 ${innerStyles(
            colour,
          )}`}
        >
          {children}
        </div>
      </div>
    </Parent>
  );
};

export default AppBar;
