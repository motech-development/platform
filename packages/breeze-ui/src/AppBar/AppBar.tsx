import React, { FC, memo, ReactNode } from 'react';
import styled, { ThemeProvider } from 'styled-components';

const AppBarInner = styled.div`
  align-items: center;
  display: flex;
  min-height: 64px;
  padding: 0 1rem;
  position: relative;
`;

interface IAppBarToolbarBase {
  colour: keyof IAppBarTheme;
}

const AppBarToolbar = styled.header<IAppBarToolbarBase>`
  ${({ colour, theme }) => `
    background-color: ${theme[colour].background};
    box-sizing: border-box;
    color: ${theme[colour].colour};
    display: flex;
    flex-direction: column;
    flex-shrink: 0;
    user-select: none;
    width: 100%;
    z-index: 1100;
  `}
`;

interface IAppBarTheme {
  [name: string]: {
    background: string;
    colour: string;
  };
}

const appBarTheme: IAppBarTheme = {
  primary: {
    background: '#161616',
    colour: '#fff',
  },
  secondary: {
    background: '#f6f9fc',
    colour: '#333',
  },
};

const AppBarBase = styled.div`
  flex-grow: 1;
`;

export interface IAppBarProps {
  children: ReactNode;
  colour?: keyof IAppBarTheme;
  element?: 'header' | 'div';
}

const AppBar: FC<IAppBarProps> = ({
  children,
  colour = 'primary',
  element = 'header',
}) => (
  <ThemeProvider theme={appBarTheme}>
    <AppBarBase>
      <AppBarToolbar as={element} colour={colour}>
        <AppBarInner>{children}</AppBarInner>
      </AppBarToolbar>
    </AppBarBase>
  </ThemeProvider>
);

export default memo(AppBar);
