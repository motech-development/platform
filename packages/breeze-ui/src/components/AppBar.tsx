import { FC, ReactNode } from 'react';
import { className, themeClass } from '../utils/className';
import TTheme from '../utils/theme';
import Element from './Element';

export type TAppBarElement = 'header' | 'div' | 'nav';

export interface IAppBarProps {
  border?: boolean;
  children: ReactNode;
  colour?: TTheme;
  element?: TAppBarElement;
  fixed?: boolean;
}

const AppBar: FC<IAppBarProps> = ({
  border = false,
  children,
  colour = 'primary',
  element = 'nav',
  fixed = false,
}) => (
  <>
    {/* @tailwind: bg-blue-900 bg-gray-900 bg-green-900 bg-red-900 bg-yellow-900 */}
    <Element
      className={className(
        fixed ? 'fixed top-0 left-0 w-full' : 'relative',
        themeClass(colour, 'bg-{theme}-900'),
      )}
      as={element}
    >
      <div className="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none">
        {/* @tailwind: border-blue-800 border-gray-800 border-green-800 border-red-800 border-yellow-800 */}
        <div
          className={className(
            border ? 'border-b' : '',
            border ? themeClass(colour, 'border-{theme}-800') : '',
          )}
        >
          <div className="flex items-center justify-between h-16 px-4 sm:px-0">
            {children}
          </div>
        </div>
      </div>
    </Element>
  </>
);

export default AppBar;
