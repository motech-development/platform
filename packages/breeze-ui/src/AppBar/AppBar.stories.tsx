import { withA11y } from '@storybook/addon-a11y';
import { boolean, select, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import styled from 'styled-components';
import BaseStyles from '../BaseStyles/BaseStyles';
import Notifications from '../Notifications/Notifications';
import TableCell from '../TableCell/TableCell';
import Typography from '../Typography/Typography';
import AppBar from './AppBar';

const Title = styled(Typography)`
  && {
    margin: 0;
  }
`;

const stories = storiesOf('AppBar', module);
const elements = {
  Div: 'div',
  Header: 'header',
};
const colours = {
  Primary: 'primary',
  Secondary: 'secondary',
};

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories
  .add('Basic app bar', () => (
    <>
      <BaseStyles />

      <AppBar
        colour={select('Colour', colours, 'primary')}
        element={select('Element', elements, 'header') as 'header' | 'div'}
        fixed={boolean('Fixed', false)}
      >
        <Title component="h1" variant="h5">
          Motech Development
        </Title>
      </AppBar>
    </>
  ))
  .add('App bar with notifications', () => (
    <>
      <BaseStyles />

      <AppBar
        colour={select('Colour', colours, 'primary')}
        element={select('Element', elements, 'header') as 'header' | 'div'}
        fixed={boolean('Fixed', false)}
      >
        <Title component="h1" variant="h5">
          Motech Development
        </Title>

        <Notifications
          alert
          items={[
            {
              message: 'This is a notification',
            },
            {
              message: 'This is another notification',
            },
          ]}
          label="Notifications"
          noResults={<div />}
          row={({ message }) => <TableCell>{message}</TableCell>}
          onClick={() => {}}
        />
      </AppBar>
    </>
  ));
