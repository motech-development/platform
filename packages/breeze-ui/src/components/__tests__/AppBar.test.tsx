import { render } from '@testing-library/react';
import AppBar from '../AppBar';

describe('AppBar', () => {
  describe('theme', () => {
    it('should render the secondary theme by default', () => {
      const { container } = render(
        <AppBar>
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the primary theme', () => {
      const { container } = render(
        <AppBar colour="primary">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the secondary theme', () => {
      const { container } = render(
        <AppBar colour="secondary">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the success theme', () => {
      const { container } = render(
        <AppBar colour="success">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the danger theme', () => {
      const { container } = render(
        <AppBar colour="danger">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render the warning theme', () => {
      const { container } = render(
        <AppBar colour="warning">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('element', () => {
    it('should render as a nav element by default', () => {
      const { container } = render(
        <AppBar>
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('nav');

      expect(element).toBeInTheDocument();
    });

    it('should render as a header element', () => {
      const { container } = render(
        <AppBar element="header">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('header');

      expect(element).toBeInTheDocument();
    });

    it('should render as a div element', () => {
      const { container } = render(
        <AppBar element="div">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('div');

      expect(element).toBeInTheDocument();
    });

    it('should render as a nav element', () => {
      const { container } = render(
        <AppBar element="nav">
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('nav');

      expect(element).toBeInTheDocument();
    });
  });

  describe('position', () => {
    it('should be relative positioned by default', () => {
      const { container } = render(
        <AppBar>
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('nav');

      expect(element).toHaveClass('relative');
    });

    it('should be fixed position', () => {
      const { container } = render(
        <AppBar fixed>
          <div data-testid="content">Hello</div>
        </AppBar>,
      );
      const element = container.querySelector('nav');

      expect(element).toHaveClass('fixed top-0 left-0 w-full');
    });
  });

  it('should render the inner content', async () => {
    const { findByTestId } = render(
      <AppBar>
        <div data-testid="content">Hello</div>
      </AppBar>,
    );

    await expect(findByTestId('content')).resolves.toBeInTheDocument();
  });
});
