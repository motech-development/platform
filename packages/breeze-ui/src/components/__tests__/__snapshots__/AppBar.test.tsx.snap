// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar theme should render the danger theme 1`] = `
<nav
  class="relative bg-red-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`AppBar theme should render the primary theme 1`] = `
<nav
  class="relative bg-blue-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`AppBar theme should render the primary theme by default 1`] = `
<nav
  class="relative bg-blue-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`AppBar theme should render the secondary theme 1`] = `
<nav
  class="relative bg-gray-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`AppBar theme should render the success theme 1`] = `
<nav
  class="relative bg-green-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`AppBar theme should render the warning theme 1`] = `
<nav
  class="relative bg-yellow-900"
>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 select-none"
  >
    <div
      class=""
    >
      <div
        class="flex items-center justify-between h-16 px-4 sm:px-0"
      >
        <div
          data-testid="content"
        >
          Hello
        </div>
      </div>
    </div>
  </div>
</nav>
`;
